contributors/design-proposals/network/grpc-cni-plugin.md

```diff
@@ -0,0 +1,268 @@
+
+Kubernetes will offer `kubernetes-cni` as a CNI driver running as a DaemonSet program, and we can have the following recommendations to simplify the deployment.
+
+To deploy an DaemonSet CNI driver with CNI plugin, it is recommended that network vendors:
+- Package kubernetes team offered `out-tree-cni-driver` and `cni-plugin-binaries` into one image.
+- Offer a yaml file of DaemonSet with informations that `out-tree-cni-driver` needed.
+    - Container path of `CNI-plugin-binaries`.
+    - Mount the configmap of network-config to container and offer it’s path.
+- Create a configmap of cni plugin config.
+
+## Example Walkthrough
+
+#### Out-tree CNI driver init
+
+1. Third-party network plugin create a configmap of network config, and put CNI plugin binaries in the right position.
+2. CNI driver read network config file and make sure can access the CNI plugin binaries.
+3. CNI driver register all gRPC server and listen for connections.
+
+#### In-tree network plugin init
+
+- Use `RemoteCNINetworkPlugin` as network plugin.
+- Call cniClient for CNI driver status.
+- Out-tree cni driver return CNI driver and network information.
+
+#### Setup pod
+
+1. In-tree cni network plugin get network namespace.
+1. Call cniClient to create pod sandbox network.
+1. Out-tree cni driver build network config and runtime config, call cni plugin binaries.
+1. Out-tree CNI driver return `PodNetworkStatus`.
+
+#### Teardown pod
+
+1. In-tree cni network plugin get network namespace.
+1. Call cniClient to delete pod sandbox network.
+1. Out-tree cni driver build network config and runtime config, call cni plugin binaries.
+1. Out-tree CNI driver return delete pod sandbox result.
+
+## Examples
+
+#### Configmap of network configure
+
+```json
+{
+  "name": "pod-network",
+  "cniVersion": "0.3.0",: [
+    {
+      "type": "ptp",
+      "ipam": {
+        "type": "host-local",
+        "subnet": "10.64.1.0/24",
+	"routes": [
+	  {
+	    "dst": "0.0.0.0/0"
+	  }
+	]
+      }
+    },
+    {
+      "type": "portmap",
+      "capabilities": {"portMappings": true},
+      "noSnat": true
+    }
+  ]
+}
+```
+
+#### DaemonSet file
+
+```json
+kind: DaemonSet
+apiVersion: extensions/v1beta1
+metadata:
+  name: kubernetes-cni
+  namespace: kube-system
+spec:
+  name: kubernetes-cni
+  namespace: kube-system
+  template:
+    spec:
+      containers:
+        - name: kubernetes-cni
+          image: xxxxxxx
+          env:
+            # The network configlist container dir
+            - name: CNI_CONF_DIR
+              value: "/etc/cni/net.d"
+            # The CNI plugin container dir
+            - name: CNI_PLUGIN_DIR
+              value: "/host/opt/cni/plugin"
+          volumeMounts:
+            - mountPath: /etc/cni/net.d
+              name: cni-conf
+              readOnly: false
+            - mountPath: /host/opt/cni/plugin
+              name: cni-plugin
+              readOnly: false
+      volumes:
+        # cni plugin binaries path
+        - hostPath:

```

Comment:
Perhaps I misunderstood some of the goals, but I thought we were trying to remove the need to have a host path volume mount in order to install CNI plugins.

Couldn't we run two containers in the pod - one with the "kubernetes-cni" image and one with the third-party CNI binaries and config, with sharing done using an emptyDir volume?